File: frontend/src/types/frozenPaperTrading.ts
export interface FrozenStrategyConfig {
  version_id: string;
  strategy_id: string;
  strategy_name: string;
  symbol: string;
  status: string;
  buy_threshold: number;
  stop_loss: number;
  target: number;
  quantity: number;
  side: string;
  use_trend_filter: boolean;
  change_percent_threshold: number;
  use_volatility_filter: boolean;
  max_adverse_volatility: number;
  allocated_capital: string | number;
  position_sizing_rule: string;
  execution_mode: string;
  same_candle_ambiguity_rule: string;
  slippage_model: string;
  transaction_cost_model: string;
  source_research_version: string;
  optimization_windows_evaluated: number;
  wfa_oos_trades_count: number;
  wfa_oos_win_rate_pct: number;
  wfa_oos_profit_factor: number;
  wfa_oos_net_pnl: string | number;
  holdout_trades_count: number;
  holdout_win_rate_pct: number;
  holdout_profit_factor: number;
  holdout_net_pnl: string | number;
  created_at_utc: string;
  config_hash: string;
}

export interface FrozenPaperPosition {
  symbol: string;
  side: string;
  quantity: string | number;
  entry_price: string | number;
  current_price: string | number;
  unrealized_pnl: string | number;
  unrealized_return_pct: number;
  entry_timestamp: string;
  stop_loss_price: string | number;
  target_price: string | number;
}

export interface FrozenPaperTradeAudit {
  trade_id: string;
  strategy_version: string;
  symbol: string;
  side: string;
  entry_timestamp: string;
  entry_price: string | number;
  exit_timestamp: string;
  exit_price: string | number;
  quantity: string | number;
  gross_pnl: string | number;
  turnover: string | number;
  brokerage: string | number;
  stt_tax: string | number;
  exchange_charges: string | number;
  stamp_duty: string | number;
  gst: string | number;
  total_charges: string | number;
  net_pnl: string | number;
  return_pct: number;
  exit_reason: string;
}

export interface FrozenPaperValidationScorecard {
  min_required_trades: number;
  completed_trades: number;
  sample_size_satisfied: boolean;
  win_rate_pct: number;
  win_rate_satisfied: boolean;
  profit_factor: number;
  profit_factor_satisfied: boolean;
  cumulative_net_pnl: string | number;
  net_pnl_positive: boolean;
  max_drawdown_pct: number;
  drawdown_acceptable: boolean;
  tp_sl_invariant_violations?: number;
  nav_accounting_discrepancy?: string | number;
  overall_status: 'INSUFFICIENT_SAMPLE_SIZE' | 'ROBUST_WINNER' | 'REJECTED_UNDERPERFORMING' | string;
  promotion_readiness_status?: 'INSUFFICIENT_SAMPLE' | 'REJECTED' | 'PAPER_VALIDATED' | 'ROBUST_PAPER_CANDIDATE' | string;
  live_promotion_blocked?: boolean;
}

export interface FrozenPaperSessionState {
  session_id: string;
  version_id: string;
  status: string;
  execution_mode: string;
  initial_capital: string | number;
  cash_balance: string | number;
  portfolio_nav: string | number;
  cumulative_realized_pnl: string | number;
  unrealized_pnl: string | number;
  total_charges_paid: string | number;
  total_trades_count: number;
  winning_trades_count: number;
  losing_trades_count: number;
  win_rate_pct: number;
  profit_factor: number;
  max_drawdown_pct: number;
  active_position: FrozenPaperPosition | null;
  validation_scorecard: FrozenPaperValidationScorecard;
  last_updated_at: string;
}

export interface FrozenPaperStepPayload {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  change_percent: number;
}

export interface FrozenPaperFeedSimulationPayload {
  candle_count?: number;
  seed?: number;
  base_volatility_pct?: number;
  drift_pct?: number;
}

File: backend/app/schemas/frozen_paper_trading.py
import uuid
from decimal import Decimal
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class FrozenStrategyConfigResponse(BaseModel):
    """Immutable metadata and parameters for a frozen strategy configuration."""
    version_id: str
    strategy_id: str
    strategy_name: str
    symbol: str
    status: str
    buy_threshold: float
    stop_loss: float
    target: float
    quantity: int
    side: str
    use_trend_filter: bool
    change_percent_threshold: float
    use_volatility_filter: bool
    max_adverse_volatility: float
    allocated_capital: Decimal
    position_sizing_rule: str
    execution_mode: str
    same_candle_ambiguity_rule: str
    slippage_model: str
    transaction_cost_model: str
    source_research_version: str
    optimization_windows_evaluated: int
    wfa_oos_trades_count: int
    wfa_oos_win_rate_pct: float
    wfa_oos_profit_factor: float
    wfa_oos_net_pnl: Decimal
    holdout_trades_count: int
    holdout_win_rate_pct: float
    holdout_profit_factor: float
    holdout_net_pnl: Decimal
    created_at_utc: str
    config_hash: str


class FrozenPaperPositionResponse(BaseModel):
    """Active open virtual position in a paper session."""
    symbol: str
    side: str  # "LONG"
    quantity: Decimal
    entry_price: Decimal
    current_price: Decimal
    unrealized_pnl: Decimal
    unrealized_return_pct: float
    entry_timestamp: str
    stop_loss_price: Decimal
    target_price: Decimal


class FrozenPaperTradeAuditResponse(BaseModel):
    """Detailed audit record of a completed virtual trade."""
    trade_id: str
    strategy_version: str
    symbol: str
    side: str
    entry_timestamp: str
    entry_price: Decimal
    exit_timestamp: str
    exit_price: Decimal
    quantity: Decimal
    gross_pnl: Decimal
    turnover: Decimal
    brokerage: Decimal
    stt_tax: Decimal
    exchange_charges: Decimal
    stamp_duty: Decimal
    gst: Decimal
    total_charges: Decimal
    net_pnl: Decimal
    return_pct: float
    exit_reason: str  # "TAKE_PROFIT", "STOP_LOSS", "MANUAL_CLOSE"


class PromotionReadinessStatus(str):
    INSUFFICIENT_SAMPLE = "INSUFFICIENT_SAMPLE"
    REJECTED = "REJECTED"
    PAPER_VALIDATED = "PAPER_VALIDATED"
    ROBUST_PAPER_CANDIDATE = "ROBUST_PAPER_CANDIDATE"


class FrozenPaperValidationScorecard(BaseModel):
    """Real-time assessment against N >= 50 trades robust quality gate and promotion readiness."""
    min_required_trades: int = 50
    completed_trades: int
    sample_size_satisfied: bool
    win_rate_pct: float
    win_rate_satisfied: bool  # >= 40%
    profit_factor: float
    profit_factor_satisfied: bool  # >= 1.10
    cumulative_net_pnl: Decimal
    net_pnl_positive: bool  # > 0
    max_drawdown_pct: float
    drawdown_acceptable: bool  # <= 15%
    tp_sl_invariant_violations: int = 0
    nav_accounting_discrepancy: Decimal = Decimal("0.0000")
    overall_status: str  # Backward-compatible status string
    promotion_readiness_status: str  # "INSUFFICIENT_SAMPLE", "REJECTED", "PAPER_VALIDATED", "ROBUST_PAPER_CANDIDATE"
    live_promotion_blocked: bool = True  # Strict safeguard: Live promotion is permanently blocked


class FrozenPaperSessionStateResponse(BaseModel):
    """Full operational snapshot of the active paper trading session."""
    session_id: str
    version_id: str
    status: str  # "ACTIVE", "PAUSED", "STOPPED"
    execution_mode: str = "PAPER"
    initial_capital: Decimal
    cash_balance: Decimal
    portfolio_nav: Decimal
    cumulative_realized_pnl: Decimal
    unrealized_pnl: Decimal
    total_charges_paid: Decimal
    total_trades_count: int
    winning_trades_count: int
    losing_trades_count: int
    win_rate_pct: float
    profit_factor: float
    max_drawdown_pct: float
    active_position: Optional[FrozenPaperPositionResponse] = None
    validation_scorecard: FrozenPaperValidationScorecard
    last_updated_at: str


class FrozenPaperStepRequest(BaseModel):
    """Single candle input for manual stepping."""
    timestamp: str
    open: float
    high: float
    low: float
    close: float
    change_percent: float


class FrozenPaperFeedSimulationRequest(BaseModel):
    """Parameters to simulate an independent market feed generating >= 30 trades."""
    candle_count: int = Field(default=80, ge=10, le=500)
    seed: int = Field(default=42, ge=1)
    base_volatility_pct: float = Field(default=1.2, ge=0.1, le=5.0)
    drift_pct: float = Field(default=0.1, ge=-2.0, le=2.0)
